http.ts

```diff
@@ -6,35 +6,46 @@ import { assert } from "./util";
 
 export async function* serve(addr: string) {
   const listener = listen("tcp", addr);
-  while (true) {
-    const c = await listener.accept();
-    yield* serveConn(c);
-  }
-  listener.close();
-}
+  // For racing promises
+  const pool: Promise<any>[] = [];
+  const untrackFromPool = (p: Promise<any>) => {
+    const requestIndex = pool.indexOf(p);
+    if (requestIndex >= 0) {
+      pool.splice(requestIndex, 1);
+    }
+  };
+  // Push conn promise also to pool to avoid starving
+  let trackedConnPromise = listener.accept().then(conn => [
+    conn,
+    untrackFromPool.bind(null, trackedConnPromise),
+  ]);
+  pool.push(trackedConnPromise);
 
-export async function* serveConn(c: Conn) {
-  let bufr = new BufReader(c);
-  let bufw = new BufWriter(c);
-  try {
-    while (true) {
-      const [req, err] = await readRequest(bufr);
-      if (err == "EOF") {
-        break;
-      }
-      if (err == "ShortWrite") {
-        console.log("ShortWrite error");
-        break;
-      }
-      if (err) {
-        throw err;
-      }
-      req.w = bufw;
+  while (true) {
+    const [reqOrConn, untrack] = await Promise.race(pool);
+    untrack(); // remove from pool
+    if (!Array.isArray(reqOrConn)) {
+      // result is not array, means from connPromise
+      // Push readRequest promise for conn
+      const trackedReqPromise = readRequest(reqOrConn)
+        .then((maybeReq) => [
+          maybeReq,
+          untrackFromPool.bind(null, trackedReqPromise)
+        ]);
+      pool.push(trackedReqPromise);
+      // Prepared to accept another connection (avoid starving)
+      trackedConnPromise = listener.accept().then(conn => [
```

Comment:
Can the logic be reorganized so there is only one call to accept?

Reply:
Just made an alternative implementation that completely got rid of `Promise.race`
(Not quite sure if there are mistakes...)